src/main/scala/com/fulcrumgenomics/umi/UmiConsensusCaller.scala

```diff
@@ -313,36 +320,33 @@ trait UmiConsensusCaller[C <: SimpleRead] {
     * NOTE: filtered out reads are sent to the [[rejectRecords]] method and do not need further handling
     */
   protected[umi] def filterToMostCommonAlignment(recs: Seq[SourceRead]): Seq[SourceRead] = {
```

Comment:
I find this new implementation impossible to follow.  I've literally been staring at it side-by-side in IntelliJ with the old implementation for ~10 minutes and I still struggle to follow it.  Two thoughts:

1. Unless this makes a very significant different I would just revert to the old method that I think was easier to follow.
2. If it does make a significant difference, then I think we need to find a clearer implementation.  I wonder if something like the following would work:

```scala
case class AlignmentGroup(cigar: Cigar, reads: mutable.Buffer[SourceRead])

protected[umi] def filterToMostCommonAlignment(recs: Seq[SourceRead]): Seq[SourceRead] = {
  val groups = new ArrayBuffer[AlignmentGroup]
  recs.sortBy(r => -r.length).foreach { rec =>
    val simpleCigar = simplifyCigar(rec.cigar)
    var found = false
    groups.foreach { g => if (simpleCigar.isPrefixOf(g.cigar) { g.reads += rec; found = true } }
    if (!found) {
      val newGroup = AlignmentGroup(simpleCigar, new ArrayBuffer[SourceRead](recs.size))
      newGroup += rec
      groups += newGroup
    }
  }

  if (groups.isEmpty) {
    Seq.empty
  }
  else {
    val sorted  = groups.sortBy(g => - g.size)
    val keepers = sorted.head
    val rejects = recs.filter(r => !keepers.contains(r))
    rejectRecords(rejects.flatMap(_.sam), FilterMinorityAlignment)
   keepers
  }
}
```

Reply:
The problem with the previous and your implementation is the `keepers.contains` method can be really really slow if we have many raw reads with the same cigar (think 20 cigar groups, each with 1000 reads).  This is the point of optimizing, and it really does make a difference. I'll try to clean it up.

Reply:
I see.  I'm curious if you tried either a) replacing the `recs.filter` with a `recs.diff(keepers)` or be creating a `val keepSet = keepers.toSet` and then calling contains on that?  You'd still pay the cost of the conversion to a set, but then the lookup time would be constant.